### PR TITLE
fix(kill): attribute a shared companion to its own profile, or to none (#772)

### PR DESCRIPTION
**Pull request by Shieldxx**
## Summary

`getProfileCompanionTargets` keyed its target map by exe **basename** and wrote `gameKey` from whichever profile the loop happened to be on. With a `gameKey` the loop is filtered to one profile so it could never collide, which is why per-row Close Apps stayed shipped. The all-profiles close walks every profile, so a utility shared across profiles (Oculus Tray Tool, SimHub, CrewChief) was overwritten and **the last profile enumerated won**. A failed close was then filed under a game the user had not been playing, and that game surfaced as having leftover apps.

This is the documented reason the tray Close Apps item was pulled hours before the 1.0.0 tag, and the blocker on re-landing it in #519.

## What changed

Targets are keyed by **identity** instead of basename: curated utilities by process name, configured companions by normalized path. Each target carries **every** profile that configures it rather than a single winner.

Attribution then follows from the data instead of from enumeration order:

| Owners | Attributed to |
|---|---|
| one | that game, exactly as before |
| several | nothing |

`undefined` there is the honest answer, not a fallback. A companion SimLauncher launched itself is attributed from `runningProcesses` and never reaches this code. What does reach it is a companion that is merely *running*, and when several profiles configure it there is no fact that makes one of them its owner. **The failure is still reported either way** — `finalizeKillAttempts` builds `failures` from the attempts alone, so the toast names the app regardless of whether anything can be attributed. What is withheld is only the persistent per-game row, which is the part users saw as a phantom.

## Two things that fell out of it

**A silently missed kill, fixed.** Two profiles pointing at same-named exes in different folders are two different processes. One Map key held them both, so the all-profiles close only ever closed the survivor. Keying paths by path closes both.

**A regression I introduced and then closed.** A curated utility whose executable path is *also* configured produced one Map key under the old scheme (the path branch enumerated second, so it won) and two under the new one. That would have issued both an `/IM` and a path-scoped kill for one app, closing it twice and reporting `closedCount` 2. The collapse is now deliberate and keeps the same winner. No existing test covered an enabled utility with a configured path, which is why the suite stayed green through it.

## `hasClosableLaunchedApps`

The issue asks whether the `KEEP IN SYNC` contract needs the same fix. It does not. It reads `target.processName` only, never `gameKey`, so attribution is irrelevant to it. The keying change can only add entries, never remove them, so its answer can move from false to true but never the reverse: a companion previously collapsed away is now correctly reported as closable, which is the direction #519 wants.

## Test plan

- [x] `format:check`, `lint`, `typecheck`, `build`
- [x] `npm test` - **596 tests**, 78 files
- [x] Five new tests: the shared-utility case, single-owner attribution, the per-game path unchanged, both same-named exes closed, and the curated-utility-with-a-path collapse
- [x] The two defect tests **fail against pre-fix `kill.ts`**; the guard tests pass on both, which is what makes them guards
- [x] Mutation verified: never-attribute, always-first-owner and always-last-owner are each caught, as are removing or inverting the collapse
- [ ] Manual: not reachable from the UI today. The path has no caller until #519 re-lands the tray item, so this is unit-level until then

Closes #772

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved process cleanup when utilities or executables share names or appear in different directories.
  * Ensured configured executable paths are respected, preventing unrelated processes from being affected.
  * Preserved accurate game attribution for uniquely owned utilities while avoiding arbitrary attribution for shared utilities.
  * Prevented duplicate cleanup attempts and phantom failures for non-running, missing, or respawning processes.
  * Improved handling of process lookup failures and cleanup across multiple profiles.

* **Tests**
  * Added regression coverage for shared names, per-game closing, path-specific matching, ownership attribution, lookup failures, and duplicate prevention.

Closes #772

---

## Update: three review findings, all of them defects this PR introduced

None were in the original bug. All three came from broadening the fix past what #772 asked for, and each is now covered.

| Finding | Source | What it was |
|---|---|---|
| The collapse matched on process name alone | CodeRabbit + review-bot | #772 reintroduced through its own fix |
| `every` could be weakened to `some` | mutation, not review | untested boundary in the collapse |
| A never-running path reported as a leftover | Codex `P1` | #772's symptom again, plus a doubled count |

**The collapse was too broad.** Deleting a name-scoped target whenever *any* path target shared its process name works for a real duplicate, since `getProfileTrackablePaths` pulls `appPaths[key]` exactly when the utility is enabled and both scopes end up with the same owners. It does not work for a basename shared by accident with another profile's freeform `trackedProcessPaths`: that deleted a target the first profile really configured, leaving the survivor looking uniquely owned by a profile that never enabled the utility, and dropping its `/IM` coverage. Review-bot's framing was the sharper one, so the condition is now a superset check rather than the owner merge CodeRabbit proposed, which would have fixed the attribution and still lost the coverage.

**Codex's P1 was the same class one layer down.** Scheduling is gated on the tasklist, which knows image names and not paths, so two same-named paths both get scheduled even when only one is running. The absent one returns 0 PIDs while the image *is* in the tasklist, which `finalizeKillAttempts` reads as an invisible elevated process (#390). Both halves were reproduced with failing tests before any code changed: a leftover registered under a profile with nothing running, and one app counted as two.

The discriminator needs no extra lookups. If another attempt in the same batch found real PIDs for that image, its presence in the tasklist is already accounted for, so a hidden instance at a *different* path cannot be inferred, and an attempt that found nothing is neither a close nor a failure. It stays narrow on purpose: with no sibling to explain the image the ambiguity is real, and the general name-versus-path discriminator is #674, which is in this milestone and blocked by this issue.

## Test plan (updated)

- [x] `format:check`, `lint`, `typecheck`, `build`
- [x] `npm test` <- **600 tests**, 78 files
- [x] Nine tests for #772. The defect tests fail against pre-fix `kill.ts`; the guard tests pass on both
- [x] Codex's two cases written as failing tests first, then fixed
- [x] Mutation verified across all three fixes, eleven mutants, all caught. Notably, making the P1 rule unconditional is caught by the existing #390 test, which is what pins it as narrow rather than blanket
- [ ] Manual: not reachable from the UI until #519 re-lands the tray item

### CODE REVIEW

File: src/main/processes/kill.ts [hunk 1]
```diff
@@ -585,16 +649,36 @@ function getProfileCompanionTargets(gameKey?: string) {
...
+  // on the same Map key and the path branch happened to be enumerated second.
+  // Keying by identity separated them, so the collapse now has to be deliberate.
+  const pathScopedNames = new Set<string>()
+  companionTargets.forEach((target) => {
+    if (target.scope === 'path') {
+      pathScopedNames.add(target.processName)
+    }
+  })
+  companionTargets.forEach((target, key) => {
+    if (target.scope === 'name' && pathScopedNames.has(target.processName)) {
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/processes/kill.ts]: Right, and this is the sharper of the two framings of it, so the fix follows yours rather than the merge that was proposed alongside it.

Collapsing now requires the path-scoped target's owners to cover **every** owner of the name-scoped one. A genuine duplicate always satisfies that: `getProfileTrackablePaths` pulls `appPaths[key]` exactly when the utility is enabled, so both scopes carry the same owner set. A coincidental basename does not, so both targets survive and the curated `/IM` coverage stays.

Merging the owners instead would have fixed the attribution and still lost that coverage, which is why superset won.

Fixed in b5a4067, plus a regression test for exactly your case: `ac` enables the utility with no path, `iracing` tracks a same-named exe elsewhere, both kills issue, and each is attributed to the profile that configured it.

One more fell out of it. Mutation testing showed `every` could be weakened to `some` with the suite still green, so nothing pinned the boundary. The case that does: one profile enables the utility **and** tracks its exe, another enables it with no path. The path target covers only the first, so collapsing would leave the second with no coverage at all. Pinned in 8545a1c.

File: src/main/processes/kill.ts [hunk 2]
```diff
@@ -585,16 +649,36 @@ function getProfileCompanionTargets(gameKey?: string) {
...
+      companionTargets.delete(key)
+    }
+  })
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/processes/kill.ts:845]: Confirmed, and fixed in b5a4067, though not by merging the owners.

You and stashpeak-review-bot found the same defect from different sides. Merging owners does fix the misattribution: the surviving path target would carry both profiles and resolve to unattributed rather than to the wrong game. But it does not fix the second half you flagged in your own note, that the curated `/IM` kill stops running and an instance at another path is no longer closed. With merged owners that narrowing is permanent.

So the collapse is now conditional instead: it only happens when the path-scoped target's owners cover **every** owner of the name-scoped one. A true duplicate satisfies that by construction, since `getProfileTrackablePaths` pulls `appPaths[key]` exactly when the utility is enabled and both scopes end up with the same owner set. A basename shared by coincidence does not, so both targets survive, nothing is misattributed, and the `/IM` coverage is kept rather than narrowed.

To your question about whether the narrowing was intended: for a real duplicate, yes, deliberately. The path-scoped kill is the precise one and closes the same process. For anything that is not a duplicate it was not intended, and no longer happens.

Regression test added as you suggested, plus one for a path target that covers only one of two owners, which mutation testing showed was the untested boundary between `every` and `some`.

File: src/main/processes/kill.ts [hunk 3]
```diff
@@ -661,7 +781,13 @@ export async function killLaunchedApps(gameKey?: string): Promise<KillResult> {
 
   companionTargets.forEach((target) => {
     if (processNames.has(target.processName)) {
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/processes/kill.ts:918]: Confirmed, both halves, and it is a hazard this PR created rather than a pre-existing one. Fixed in e7d0f03.

I wrote the two tests before touching the code, and they failed exactly as described:

- real instance still running: `failedCount` 2, with a leftover registered under the profile whose path had nothing on it. That is #772's own symptom, reintroduced by its fix.
- real instance closes: `closedCount` 2 for one app, because the image leaving the tasklist reads as success for every attempt including the empty one.

**The fix, and why it is not a new WMI pass.** Verifying each path before scheduling would either cost a lookup per target or, if the answer were taken as authoritative, regress #390: 0 PIDs at a configured path is exactly what an elevated process with a null `ExecutablePath` looks like, so absence cannot be trusted on its own.

The discriminator is already in the batch. If another attempt found real PIDs for that image name, the image's presence in the post-kill tasklist is **already accounted for**, so inferring a hidden instance at a *different* path is unjustified. Such an attempt is then neither a failure nor a close: it closed nothing.

Deliberately narrow. With no sibling to explain the image the ambiguity is real and stays unresolved, because discriminating name from path in the general case is #674, which is already in this milestone and blocked by this issue.

Mutation verified, all caught: removing the guard brings the false leftover back, removing the count adjustment brings the double count back, and making the predicate unconditional is caught by the existing #390 test, which is what pins the rule as narrow rather than blanket.

File: src/main/processes/kill.ts [hunk 4]
```diff
@@ -416,6 +416,26 @@ export async function finalizeKillAttempts(
...
+  // did find PIDs, so inferring a hidden instance at a DIFFERENT path is
+  // unjustified: it invented a leftover under a profile that had nothing
+  // running, and counted an empty attempt as a closed app (Codex P1 on #818).
+  //
+  // Deliberately narrow. With no sibling to explain the image, the ambiguity is
+  // real and stays unresolved; discriminating name from path in the general case
+  // is #674.
+  const imagesWithConfirmedTarget = new Set(
+    attempts.filter((attempt) => attempt.notFound !== true).map((attempt) => attempt.processName)
+  )
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/processes/kill.ts]: Confirmed and fixed in 87a0fe7. `notFound !== true` was the wrong test: the lookup-error branch returns neither flag, so a WMI call that blew up was vouching for a sibling it never looked at.

Attempts now carry `targetConfirmed`, set only on positive evidence that a process under that image name was actually observed:

- PIDs discovered at the path
- a tracked child we held a real PID for, unless taskkill says it had already exited
- an `/IM` taskkill that killed something or was denied by something

Ambiguous failures, including the lookup error, are explicitly **not** evidence. Conservative on purpose: less confirmation means the elevated inference is preserved more often, which is the safe direction.

Your regression test is in, with a lookup-error fixture the mock never had. It arms ac's path to fail the WMI call while iracing's looks elevated-invisible, and asserts iracing still lands in `unclosedProcesses` as `access_denied`.

**Wiring it up surfaced a second defect, which is the more interesting half.** `targetConfirmed` and `notFound` can both hold on one attempt: PIDs get discovered, then taskkill finds them already gone. With a Set, such an attempt appeared in `imagesWithConfirmedTarget` and **vouched for itself**, so it classified itself as empty. Three existing tests went red, which is the only reason it did not ship. It is now a count per image name minus self, so the rule genuinely requires another attempt.

**One residual the rule still cannot see**, documented in place and recorded on #674 rather than left implicit: a sibling that found its target and closed it successfully still counts as confirmation, even though the image surviving afterwards can only be something else, possibly a real elevated-invisible process at the path being judged. Separating those needs each sibling's post-kill state, which is #674's discriminator. The trade is deliberate, since the false positive suppressed is #772's actual symptom and far likelier than the false negative left.

Mutation verified: reverting to `notFound !== true`, dropping the self-exclusion, and marking the lookup-error branch confirmed are all caught.

File: src/main/processes/kill.ts [hunk 5]
```diff
@@ -440,6 +498,14 @@ export async function finalizeKillAttempts(
       const appPath = attempt.appPath
       const isFullPathAttempt = isFullExePath(appPath)
 
+      // This attempt found nothing at its own path, and another attempt in the
+      // same batch did find the image elsewhere. So this path was never running:
+      // not a failure to close, and not a close either.
+      const siblingsWithConfirmedTarget =
+        (confirmedTargetsByImage.get(attempt.processName) ?? 0) - (attempt.targetConfirmed ? 1 : 0)
+      const isEmptySameNameTarget =
+        isFullPathAttempt && attempt.notFound === true && siblingsWithConfirmedTarget > 0
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/processes/kill.ts]: Confirmed, fixed in ffe6304.

The `/IM` confirmation was doing the one thing it cannot do: it establishes that *some* process under that image name exists, and the process it found may be exactly the one the other profile tracks by path. Using it to declare that path empty then suppressed the tracked profile's leftover and pinned the failure on the curated profile alone, which is the misattribution this PR exists to remove.

Confirmation now comes only from a **path-scoped** sibling at a **different** path. That also replaced the previous count-minus-self guard: comparing normalized paths says what it means, whereas subtracting one's own contribution from a per-image count was an oblique way of saying the same thing and only covered the self case, not yours.

File: src/main/processes/kill.ts [hunk 6]
```diff
@@ -507,7 +575,13 @@ export async function finalizeKillAttempts(
...
       (!attempt.success && !attempt.notFound && !attempt.imageGoneFromTasklist)
   )
-  const closedCount = finalizedAttempts.length - failedAttempts.length
+  // An attempt that found nothing at its own path while a sibling accounted for
+  // the image closed nothing, so it is neither a success nor a failure. Counting
+  // it as closed reported one app as two.
+  const emptySameNameTargets = finalizedAttempts.filter(
+    (attempt) => attempt.isEmptySameNameTarget
+  ).length
+  const closedCount = finalizedAttempts.length - failedAttempts.length - emptySameNameTargets
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/processes/kill.ts]: Confirmed, fixed in the same commit.

`isEmptySameNameTarget` is now also gated on the post-kill lookup returning 0 PIDs, not just the pre-kill one. That makes empty and failed **mutually exclusive by construction**: with no PIDs on the recheck, none of `stillRunning`'s disjuncts can hold, so such an attempt can never reach `failedAttempts` and can never be subtracted twice.

I went that way rather than filtering the empty count by `!stillRunning` because the gate is also the honest statement. An attempt with PIDs on the recheck is not an empty target that happens to have failed; it is a target that was there, which is what your respawn case describes.

Regression test in: ac is a real confirmed sibling that fails to close, iracing reports nothing pre-kill and is back by the recheck, and the result is `closedCount` 0 with `failedCount` 2.

That needed a per-path WMI lookup counter in the test mock. The existing one is keyed by process name, which two same-named paths share, so it could not express "absent on the first lookup for this path" and my first attempt at the test was silently measuring the wrong thing.

Mutation verified, all caught: letting bare-name confirmations vouch again, letting a same-path sibling vouch, dropping the post-kill emptiness gate, and dropping the count adjustment.

File: src/main/processes/kill.ts [hunk 7]
```diff
@@ -442,14 +510,34 @@ export async function finalizeKillAttempts(
...
       if (isFullPathAttempt) {
         const { processIds } = await findProcessIdsByExecutablePath(attempt.processName, appPath)
+
+        const ownPath = normalizePathForComparison(appPath)
+        const confirmedPaths = confirmedPathsByImage.get(attempt.processName)
+        isEmptySameNameTarget =
+          attempt.notFound === true &&
+          processIds.length === 0 &&
+          !!confirmedPaths &&
+          Array.from(confirmedPaths).some((confirmedPath) => confirmedPath !== ownPath)
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/processes/kill.ts:567]: Confirmed, fixed in f6b1719, and this is the same shape you caught one lookup earlier: an absent signal read as information.

`findProcessIdsByExecutablePath` resolves `{ processIds: [], detail }` on failure, so a zero count means either "nothing is there" or "could not look". I gated the pre-kill side on `targetConfirmed` after your last finding and then made the identical assumption on the post-kill recheck. It now requires the recheck to have actually succeeded.

Regression test in, with the post-kill lookup failure fixture the mock did not have: ac is a genuinely confirmed sibling at a different path that fails to close, iracing finds nothing pre-kill and its recheck blows up, and iracing's leftover is still reported as `access_denied`. Dropping the gate turns it red.

**Since this was the second instance, I checked the rest of the file for a third.** Two callers only:

- `killProcessByImageName` (`:284`) returns on the `detail` branch **before** its `processIds.length === 0` check, so a failed lookup never reaches it. Correct already.
- the `stillRunning` disjunct at `:555` reads `processIds.length > 0` as positive evidence, so a failed lookup contributes false and the attempt falls through to `isElevatedInconclusive`, which for a notFound attempt with the image present is true. It fails toward reporting the leftover rather than hiding it, which is the safe direction.

So no third instance. The hazard was specific to the two places I added.

File: src/main/processes/kill.ts [hunk 8]
```diff
@@ -507,7 +604,13 @@ export async function finalizeKillAttempts(
...
       (!attempt.success && !attempt.notFound && !attempt.imageGoneFromTasklist)
   )
-  const closedCount = finalizedAttempts.length - failedAttempts.length
+  // An attempt that found nothing at its own path while a sibling accounted for
+  // the image closed nothing, so it is neither a success nor a failure. Counting
+  // it as closed reported one app as two.
+  const emptySameNameTargets = finalizedAttempts.filter(
+    (attempt) => attempt.isEmptySameNameTarget
+  ).length
+  const closedCount = finalizedAttempts.length - failedAttempts.length - emptySameNameTargets
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply src/main/processes/kill.ts]: Confirmed, fixed across a373342 and the follow-up.

Counting and emptiness are now separate questions with separate evidence, which is what the case needed:

- **closed nothing** accepts a successful bare-name `/IM` sibling. That kill is not path-scoped, so it takes down whatever ran under the name including what the path target was aiming at, and the path attempt then finds nothing.
- **the path was empty** still requires a confirmed sibling at a different path. A successful `/IM` cannot prove it, because the process it killed may be exactly the one that path tracks. That is your earlier P2 on this PR and it stays intact.

The count is also derived by filtering rather than by subtracting two overlapping tallies, so an attempt cannot be deducted twice and the total cannot go negative.

**Two things worth reporting rather than just claiming a fix.**

Dropping the sibling requirement from the count entirely was cleaner and I tried it first. It turned five existing tests red, because a lone empty attempt has long been counted as closed and changing that is a separate behaviour question, out of scope here. So a lone empty attempt keeps its existing treatment.

And the first version of this fix was only half pinned. Mutation showed that letting a successful `/IM` also prove a path empty, reopening your earlier P2, left the suite green: the existing guard uses an access-denied `/IM`, so it never reaches the success branch. The discriminating case is a `/IM` that succeeds while an elevated instance at the other profile's path survives. Added, along with the fixture for it, since the mock deleted the image name on every successful `/IM`.

All four mutants caught now: dropping the subsumption, letting `/IM` prove emptiness, ignoring `closedNothing` in the count, and letting path-scoped attempts subsume by name.
